proto/stream_service.proto

```diff
@@ -25,8 +25,12 @@ message BroadcastActorInfoTableRequest {
 
 // Create channels and gRPC connections for a fragment
 message BuildActorsRequest {
+  message SubscriptionIds {
+    repeated uint32 subscription_ids = 1;
+  }
   string request_id = 1;
   repeated uint32 actor_id = 2;
+  map<uint32, SubscriptionIds> related_subscriptions = 3;
```

Comment:
Placing such a field here (and all the related code path) looks quite ad-hoc to me... Why isn't it a field in `stream_plan::StreamActor`? 

A related question: how is this information persisted?

Reply:
We just use the subscription information persisted in catalog.

I used to think about adding it as a field of `StreamActor`. But since `StreamActor` is also persisted, we will have to carefully maintain the consistency between the subscription information persisted in catalog and the `StreamActor`. However, such information in `StreamActor` is redundant, and the effort to carefully maintain the consistency is unnecessary. 

So at the end I chose the current implementation.

Plus: I think the dispatcher information is also unnecessary to be persisted in `StreamActor`, since, IIUC, the dispatcher information can be recalculated when we know the upstream and downstream distribution, and therefore the persisted dispatcher information is also unnecessary, and we will also have to carefully maintain the consistency.

Reply:
> We just use the subscription information persisted in catalog.

This sounds ok to me. 

My concern is on this field `related_subscriptions`. Basically, the `BuildActorsRequest` RPC and all the subsequent calls only pass `actor_ids`, while the additional information such as the definition of the actors are stored in metadata and collected later. To keep the code consistent and clean, I think here we should also follow this style.


Plus:

> I think the dispatcher information is also unnecessary to be persisted in StreamActor

The actors need to known exactly which downstream actors it should connect to. Imagine that when one CN fails and restarts, the actors on it will be recreated while the rest CN doesn't need to do so. Does anywhere preserve such information?

Reply:
> The actors need to know exactly which downstream actors it should connect to. Imagine that when one CN fails and restarts, the actors on it will be recreated while the rest CN doesn't need to do so. Is such information preserve somewhere?

IIUC, the dispatch information can be calculated as long as we have the vnode distribution of upstream and downstream. For example, the upstream actor owns vnode [1, 2, 3], and then if we know the downstream actor A owns vnode1, B owns vnode2 and C owns vnode3, and then we will know that the upstream actor should dispatch to downstream actor A, B and C. 

Currently, when a CN fails, a recovery will be triggered, and all actors will be rebuilt. Meta node should have the persisted vnode distribution information of all actors, and then the dispatcher information can be calculated as well, and then the actor can know which actors to connect.

Reply:
>  For example, the upstream actor owns vnode [1, 2, 3], and then if we know the downstream actor A owns vnode1, B owns vnode2 and C owns vnode3, and then we will know that the upstream actor should dispatch to downstream actor A, B and C.

I see. You mean the `vnode_bitmap` in the downstream actors, which defines the data distribution of all state tables in that fragment/actors. Technically, it seems possible to have different distribution key in shuffle and state table, even though I believe in our system they are always same. 🤔